File: Java/DFS/Letter_Combinations_of_a_Phone_Number.java
// https://leetcode.com/problems/letter-combinations-of-a-phone-number/

class Solution {
    public List<String> letterCombinations(String digits) {
        int n = digits.length();
        LinkedList<String> ret = new LinkedList<>();
        if (n == 0) return ret;
        ret.add("");
        String[] tmp = new String[]{"0", "1", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        for (int i = 0; i < n; i++) {
            int m = ret.size(), j = 0;
            int k = digits.charAt(i) - '0';
            while (j < m) {
                String curr = ret.remove();
                for (char c: tmp[k].toCharArray()) {
                    ret.add(curr + c);
                }
                j++;
            }
        }
        return ret;
    }
}

File: Python/DFS/Letter_Combinations_of_a_Phone_Number.py
# https://leetcode.com/problems/letter-combinations-of-a-phone-number/

class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if digits == '':
            return []
        res = []
        dic = {'2':'abc', '3':'def', '4':'ghi', '5':'jkl', '6':'mno', '7':'pqrs', '8':'tuv', '9':'wxyz'}
        def dfs(i, ans):
            if i == len(digits):
                res.append(ans)
            else:
                for char in dic[digits[i]]:
                    ans = ans + char
                    dfs(i + 1, ans)
                    ans = ans[:-1]
        dfs(0, '')
        return res
